src/game/logic/object/weapon.cpp

```diff
@@ -1195,15 +1370,236 @@ bool Weapon::Compute_Approach_Target(const Object *source_obj,
     float angle_offset,
     Coord3D &approach_target_pos) const
 {
-    // todo
-    return false;
+    const Coord3D *pos;
+    Coord3D vector;
+
+    if (target_obj != nullptr) {
+        pos = target_obj->Get_Position();
+        g_thePartitionManager->Get_Vector_To(target_obj, source_obj, FROM_BOUNDINGSPHERE_2D, vector);
+    } else if (target_pos != nullptr) {
+        pos = target_pos;
+        g_thePartitionManager->Get_Vector_To(source_obj, target_pos, FROM_BOUNDINGSPHERE_2D, vector);
+        vector.x = -vector.x;
+        vector.y = -vector.y;
+        vector.z = -vector.z;
+    } else {
+        captainslog_dbgassert(false, "error");
+        approach_target_pos.Zero();
+        return false;
+    }
+
+    float length = vector.Length();
+    float min_range = m_template->Get_Minimum_Attack_Range();
+
+    if (min_range <= 10.0f || length >= min_range) {
+        if (length >= 0.001f) {
+            if (Is_Contact_Weapon()) {
+                approach_target_pos = *pos;
+                return false;
+            } else {
+                vector.x = vector.x / length;
+                vector.y = vector.y / length;
+                vector.z = vector.z / length;
+
+                if (angle_offset != 0.0f) {
+                    float angle = GameMath::Atan2(vector.y, vector.x);
+                    vector.x = GameMath::Cos(angle + angle_offset);
+                    vector.y = GameMath::Sin(angle + angle_offset);
+                }
+
+                float range = Get_Attack_Range(source_obj) * 0.9f;
+                approach_target_pos.x = range * vector.x + pos->x;
+                approach_target_pos.y = range * vector.y + pos->y;
+                approach_target_pos.z = range * vector.z + pos->z;
+
+                if (source_obj->Get_AI_Update_Interface() != nullptr) {
+                    const AIUpdateInterface *update = source_obj->Get_AI_Update_Interface();
+
+                    if (update->Is_Aircraft_That_Adjusts_Destination()) {
+                        g_theAI->Get_Pathfinder()->Adjust_Target_Destination(
+                            source_obj, target_obj, target_pos, this, &approach_target_pos);
+                    }
+                }
+
+                return false;
```

Comment:
`return false` is correct? Would have expected `true`, because `approach_target_pos` is set.

Reply:
Yes, verified that all the returns in this function are correct.